client/ext/bigquery/current-timestamp-optimizer_test.js

```diff
@@ -0,0 +1,266 @@
+          dashboardSetting.enabled = true;
+          widgetSetting.enabled = false;
+          
+          expect(testOptimizer.canApply(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(false);
+        });
+      });
+      
+      describe('should return true when', function() {
+        it('dashboard..enabled is true and widget..enabled is null', function() {
+          dashboardSetting.enabled = true;
+          widgetSetting.enabled = null;
+          
+          expect(testOptimizer.canApply(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(true);
+        });
+
+        it('dashboard..enabled is true and widget..enabled is true', function() {
+          dashboardSetting.enabled = true;
+          widgetSetting.enabled = true;
+          
+          expect(testOptimizer.canApply(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(true);
+        });
+
+        it('dashboard..enabled is false and widget..enabled is true', function() {
+          dashboardSetting.enabled = false;
+          widgetSetting.enabled = true;
+          
+          expect(testOptimizer.canApply(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(true);
+        });
+      });
+    });
+
+    describe('getEffectiveGranularity', function() {
+      it('should return the dashboard setting if the widget is null', function() {
+        dashboardSetting.granularity = CurrentTimestampGranularity.YEAR;
+        
+        expect(testOptimizer.getEffectiveGranularity(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(
+            CurrentTimestampGranularity.YEAR);
+      });
+
+      it('should return the widget setting if the widget is specified', function() {
+        dashboardSetting.granularity = CurrentTimestampGranularity.YEAR;
+        widgetSetting.granularity = CurrentTimestampGranularity.DAY;
+        
+        expect(testOptimizer.getEffectiveGranularity(PROVIDED_DASHBOARD, PROVIDED_WIDGET)).toEqual(
+            CurrentTimestampGranularity.DAY);
+      });
+    });
+
+    describe('apply', function() {
+      it('should leave the query untouched when canApply returns false', function() {
+        spyOn(testOptimizer, 'canApply').and.returnValue(false);
+
+        const PROVIDED_SQL = 'SELECT CURRENT_TIMESTAMP() FROM foo';
+        const EXPECTED_SQL = PROVIDED_SQL;
+        PROVIDED_WIDGET.datasource.query_exec = PROVIDED_SQL;
+
+        testOptimizer.apply(PROVIDED_DASHBOARD, PROVIDED_WIDGET);        
+        expect(PROVIDED_WIDGET.datasource.query_exec).toEqual(EXPECTED_SQL);
+      });
+
+      it('should replace CURRENT_TIMESTAMP() with the current effective date', function() {
+        const effectiveDate = new Date(PROVIDED_YEAR, PROVIDED_MONTH, PROVIDED_DAY, PROVIDED_HOUR, 0);
+        const expectedDateString = '2006-02-15T12:00:00.000Z'
+        spyOn(effectiveDate, 'toISOString').and.returnValue(expectedDateString);
+
+        spyOn(testOptimizer, 'canApply').and.returnValue(true);
+        spyOn(testOptimizer, 'getEffectiveDate').and.returnValue(effectiveDate);
```

Comment:
The test verifies that the apply() method replaces CURRENT_TIMESTAMP() within a given string.   I'm mocking out the stuff that determines what date it's replaced with, and whether or not it should return.

Based on other optimizations to the implementation, this test is clearer (and a duplicate test was removed).